hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieBaseParquetWriter.java

```diff
@@ -65,14 +63,18 @@ public HoodieBaseParquetWriter(Path file,
   }
 
   public boolean canWrite() {
-    // TODO we can actually do evaluation more accurately:
-    //      if we cache last data size check, since we account for how many records
-    //      were written we can accurately project avg record size, and therefore
-    //      estimate how many more records we can write before cut off
-    if (lastCachedDataSize == -1 || getWrittenRecordCount() % WRITTEN_RECORDS_THRESHOLD_FOR_FILE_SIZE_CHECK == 0) {
-      lastCachedDataSize = getDataSize();
+    if (getWrittenRecordCount() >= recordNumForNextCheck) {
+      long dataSize = getDataSize();
+      long avgRecordSize = dataSize / getWrittenRecordCount();
+      // Follow the parquet block size check logic here, return false
+      // if it is within ~2 records of the limit
+      if (dataSize > (maxFileSize - avgRecordSize * 2)) {
+        return false;
```

Comment:
> Follow the parquet block size check logic here

What block size check logic did you mean? Can we memo the avgRecordSize though？

Reply:
Follow this:

Reply:
Btw, I think we'd better to let `ParquetRowDataWriter` to extend `HoodieBaseParquetWriter`, so we can reduce many duplicates codes and reuse the logic check here?

Reply:
>Btw, I think we'd better to let ParquetRowDataWriter to extend HoodieBaseParquetWriter, so we can reduce many duplicates codes and reuse the logic check here?

Hi, @danny0405 do we need to refine this? If so, I'm fine to involve that in this pr.

Reply:
The strategy seems be introudced by Ryan in https://github.com/apache/parquet-mr/commit/412ab9669810921d04f9feabfbeafa906d4de506, which is used for row group alignment, do we need it here for Hoodie?

Reply:
Yes, from my understanding, if the user set the max file size, we'd better don't exceed this value.